### PR TITLE
Wayland: rework the event loop & expose readiness signal

### PR DESCRIPTION
**Pull request by vberger**
This PR does a few things:

- cleanup the wayland implementation of event loop, by not creating a 2nd wayland event queue (it is no longer needed now that each event loop has its own wayland connection)
- buffer the events generated by the wayland dispatch rather than calling the user callback on the go. This has two advantages:
   - it simplifies the events loop code
   - it allows to call Window methods such as `set_inner_size` and `set_title` from within a callback without deadlocking
- add an interface to query readiness of the window for drawing. Indeed, it is illegal to draw on a window before it has been configured by the server with an appropriate event when using xdg_shell. Doing so causes the connection to be terminated by the server. This exposes an `Arc<AtomicBool>` that becomes `true` once the appropriate event has been received.

The idea of this last change is that glutin can retrieve this `Arc<AtomicBool>` and silently ignore calls to `swap_buffers()` as long as this value is still false. Without this, all the glutin examples crash on wayland because the window was drawn on without being configured.

The rationale for this `Arc<AtomicBool>` rather than a simple method returning a `bool` to query the state is that the glutin context does not keep a reference to the winit `Window`, and as such would not be able to query for readiness. If you are okay with this design, I'll make a sibling PR to glutin for the appropriate changes.

The last commit also fixes a few errors in wayland protocol handling by winit found while testing this last changes.

### CODE REVIEW

**Comment by tomaka** [comment]
What about adding `fn is_ready(&self) -> bool` instead of `fn ready_signal(&self) -> AtomicBool`?
`is_ready` would be called instead of querying the `AtomicBool`.

**Comment by vberger** [comment]
That was my initial idea, but then, when trying to update glutin to use that, I realised that glutin's `Context` does not keep a reference to the `Window`. As such, glutin could not use the `is_ready()` API without some pretty large changes.

But if this would be the preferred way, I'll change the PR to that.

**Comment by vberger** [comment]
So, what is you opinion of this @tomaka : is it possible to have glutin keep a reference to the window to check if the surface is ready?

**Comment by tomaka** [comment]
I think the correct thing to do is to make each backend have ownership of the `Window`.

Alternatively, a hack could be to pass the `&Window` to the internal `swap_buffer` methods of the backends.

**Comment by vberger** [comment]
Okay, I'll change this PR to expose an `is_ready(..)` method instead.

**Comment by vberger** [comment]
Done :)

File: src/os/unix.rs [hunk 1]
```diff
@@ -112,6 +112,17 @@ pub trait WindowExt {
     ///
     /// The pointer will become invalid when the glutin `Window` is destroyed.
     fn get_wayland_display(&self) -> Option<*mut libc::c_void>;
+
+    /// Check if the window is ready for drawing
+    ///
+    /// On wayland, drawing on a surface before the server has configured
+    /// it using a special event is illegal. As a result, you should wait
+    /// until this methid returns `true`.
```
Comment (tomaka) [inline src/os/unix.rs]: methid -> method

**Pull request merged**